design/01-proc-macro.md

```diff
@@ -0,0 +1,290 @@
+    let a = Input { value: 1 };
+    let b = Input { value: 2 };
+    add!(a.to_value(), b.to_value());
+}
+```
+
+### Procedural macro author perspective
+
+To implement a procedural macro, a programmer will have to:
+- Create a new package, with a `Scarb.toml` manifest file and a `src/` directory besides.
+- The manifest file must define a `cairo-plugin` target type.
+- Write a Rust library, inside the `src/` directory, that implements the procedural macro API.
+- Note, that the package does not contain any `Cargo.toml` file, as it will be generated by Scarb.
+- A Rust crate exposing an API for writing procedural macros will be published for programmers. This crate will be automatically added to generated `Cargo.toml` file.
+- The Rust library contained in the package will have to implement a public `generate_code` function.
+- This function accepts a `TokenStream`  as an input and returns a `ProcMacroResult` as an output, both defined in the helper library.
+- The result enum will contain one of three values:
+    - `ProcMacroResult::Leave` - procedural macro does not take any action
+    - `ProcMacroResult::Replace({ TokenStream, AuxData })` - procedural macro has generated code that replaces input
+    - `ProcMacroResult::Remove` - original code should be removed
+    - `ProcMacroResult::Bail(ProcMacroDiagnostics)` - procedural macro exited with an error
+- Alongside the new TokenStream, a procedural macro can emit auxiliary data, encoded as an arbitrary JSON.
+
+A `TokenStream` definition:
+- A token can be defined as a string with one of three kinds assigned to it during parsing.
+- The token kinds are: identifier, punctuation symbol and literal.
+- Kinds of different tokens can be represented as an enum called `TokenTree`.
+- An iterator over token trees is called a `TokenStream`. A `TokenStream` is a representation of a Cairo code.
+- The `TokenStream` are plain Cairo tokens that will be parsed by the procedural macro implementation itself.
+
+Example:
+
+```toml
+[package]
+name = "some-macro"
+version = "0.1.0"
+
+[cairo-plugin]
+```
+
+```rust
+/// The entry point of procedural macro implementation.
+pub fn generate_code(token_stream: TokenStream) -> ProcMacroResult {
+    // no-op
+    ProcMacroResult::Leave
+}
+```
+
+## Reference-level explanation
+
+### Procedural macros will be compiled into shared libraries
+
+- Procedural macros will be implemented in Rust.
+- Procedural macros will be compiled into shared libraries, that Scarb can load at the runtime through dynamic loading mechanism of the host operating system.
+- Procedural macros will be compiled to `.dylib` files on OSX, `.so` files on Linux and `.dll` on Windows.
+- Scarb will use [`libloading` crate](https://docs.rs/libloading/latest/libloading/), which exposes a cross-platform interface for interactions with platform dynamic library loading primitives, to load shared libraries into memory.
+
+### Procedural macros will be special Scarb packages containing Rust code
```

Comment:
Two questions to resolve here:
1. Should the auto-generated Cargo.toml be included in package archives? Or should Scarb always re-generate Cargo.toml before compiling. Both variants have pros and cons I believe.
2. How will we control what version of the `cairo-lang-macro` crate will be used? How users will control compatibility with Scarb releases?

Reply:
1. Personally, I think it should be regenerated on each project resolution. This will avoid problem of compatibility of the file content between Scarb versions. 
2. One idea I have is to treat the cairo-lang-macro versioning similar to what we do with `starknet` package today. That is, use a version equal to Scarb version.

Reply:
Ad 1. 👍🏻 
Ad 2. So you mean that users will use `cairo-version` to determine what `cairo-lang-macro` version are they targeting? I wonder if this wouldn't be too restrictive for us, and it wouldn't be better to decouple versioning schemes for these; like Rust's `syn`/`quote` crates do

Reply:
Updated with information about separate versioning of interface crates